tools/ci/java-ci-tools/src/main/java/org/apache/flink/tools/ci/licensecheck/LicenseChecker.java

```diff
@@ -0,0 +1,333 @@
+			if (noticeDependencyMatcher.find()) {
+				String groupId = noticeDependencyMatcher.group(1);
+				String artifactId = noticeDependencyMatcher.group(2);
+				String version = noticeDependencyMatcher.group(3);
+				if (groupId == null && artifactId == null && version == null) { // "bundles" case
+					groupId = noticeDependencyMatcher.group(5);
+					artifactId = noticeDependencyMatcher.group(6);
+					version = noticeDependencyMatcher.group(7);
+				}
+				IncludedDependency toAdd = IncludedDependency.create(groupId, artifactId, version);
+				if (!declaredDependencies.add(toAdd)) {
+					LOG.error("Dependency {} has been declared twice in module {}", toAdd, moduleName);
+					severeIssueCount++;
+				}
+			}
+		}
+		// print all dependencies missing from NOTICE file
+		Collection<IncludedDependency> expectedDependencies = modulesWithShadedDependencies.get(moduleName);
+		for (IncludedDependency expectedDependency : expectedDependencies) {
+			if (!declaredDependencies.contains(expectedDependency)) {
+				LOG.error("Could not find dependency {} in NOTICE file {}", expectedDependency, noticeFile);
+				severeIssueCount++;
+			}
+		}
+
+		boolean moduleDefinesExcessDependencies = MODULES_DEFINING_EXCESS_DEPENDENCIES.contains(moduleName);
+
+		// print all dependencies defined in NOTICE file, which were not expected
+		Set<IncludedDependency> excessDependencies = new HashSet<>(declaredDependencies);
+		excessDependencies.removeAll(modulesWithShadedDependencies.get(moduleName));
+		for (IncludedDependency declaredDependency : declaredDependencies) {
+			if (!expectedDependencies.contains(declaredDependency)) {
+				if (moduleDefinesExcessDependencies) {
+					LOG.debug(
+						"Dependency {} is mentioned in NOTICE file {}, but was not mentioned by the build output as a bundled dependency",
+						declaredDependency,
+						noticeFile);
+				} else {
+					LOG.warn(
+						"Dependency {} is mentioned in NOTICE file {}, but is not expected there",
+						declaredDependency,
+						noticeFile);
+				}
+
+			}
+		}
+
+		return severeIssueCount;
+	}
+
+	private static List<Path> findNoticeFiles(Path root) throws IOException {
+		return Files.walk(root)
+			.filter(file -> {
+				int nameCount = file.getNameCount();
+				return file.getName(nameCount - 3).toString().equals("resources")
+					&& file.getName(nameCount - 2).toString().equals("META-INF")
+					&& file.getName(nameCount - 1).toString().equals("NOTICE");
+			})
+			.collect(Collectors.toList());
+	}
+
+	private static Multimap<String, IncludedDependency> parseModulesFromBuildResult(File buildResult) throws IOException {
+		Multimap<String, IncludedDependency> result = ArrayListMultimap.create();
+
+		try (BufferedReader br = new BufferedReader(new FileReader(buildResult))) {
+			String line;
+			String currentShadeModule = null;
+			String currentDependencyCopyModule = null;
+			while ((line = br.readLine()) != null) {
```

Comment:
```suggestion
		try (Stream<String> lines = Files.lines(buildResult.toPath()) {
			String currentShadeModule = null;
			String currentDependencyCopyModule = null;
			for (String line : lines) {
```

Reply:
In this case I have to use `for (String line : (Iterable<String>) lines::iterator) {`.

Reply:
I have applied this suggestion anyways